src/current/v23.2/read-committed.md

```diff
@@ -0,0 +1,953 @@
+
+Start a transaction:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
+~~~
+
+Check to make sure that another doctor is on call for `2023-12-05`. Use [`FOR SHARE`]({% link {{ page.version.version }}/select-for-update.md %}) to lock the rows so that they cannot be updated by another transaction:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+SELECT * FROM schedules
+  WHERE day = '2023-12-05'
+  FOR SHARE;
+~~~
+
+~~~
+     day     | doctor_id | on_call
+-------------+-----------+----------
+  2023-12-05 |         1 |    t
+  2023-12-05 |         2 |    t
+~~~
+</div>
+
+  <div class="grid-item">
+Around the same time, Doctor 2, Betty, starts to request leave for the same day using the hospital's schedule management application. 
+
+In a new terminal (Session 2), open the SQL shell on your [`cockroach demo`]({% link {{ page.version.version }}/cockroach-demo.md %}) cluster. Start a transaction:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
+~~~
+
+Check to make sure that another doctor is on call for `2023-12-05`. Use `FOR SHARE` to lock the rows so that they cannot be updated by another transaction:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+SELECT * FROM schedules
+  WHERE day = '2023-12-05'
+  FOR SHARE;
+~~~
+
+~~~
+     day     | doctor_id | on_call
+-------------+-----------+----------
+  2023-12-05 |         1 |    t
+  2023-12-05 |         2 |    t
+~~~
+  </div>
+
+  <div class="grid-item">
+In Session 1, the previous read confirmed that another doctor is available on `2023-12-05`. Update the schedule to put Abe on leave:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+UPDATE schedules SET on_call = false
+  WHERE day = '2023-12-05'
+  AND doctor_id = 1;
+~~~
+
+However, because Session 2 has also acquired a shared lock on these rows, the current transaction is blocked until Session 2 releases its lock.
+
+{{site.data.alerts.callout_success}}
+`FOR SHARE` is not typically used by transactions that intend to update rows, as shown by the following steps. For more information, see [When to use locking reads](#when-to-use-locking-reads).
+{{site.data.alerts.end}}
+  </div>
+
+  <div class="grid-item">
+In Session 2, the previous read confirmed that another doctor is available on `2023-12-05`. Update the schedule to put Betty on leave:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+UPDATE schedules SET on_call = false
+  WHERE day = '2023-12-05'
+  AND doctor_id = 2;
+~~~
+
+However, because Session 1 has also acquired a shared lock on these rows, the current transaction is blocked until Session 1 releases its lock.
+
+Because each concurrent transaction is waiting on the other to release its lock, they will deadlock. Session 2 aborts first with a [`ABORT_REASON_PUSHER_ABORTED`]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#abort_reason_pusher_aborted) error:
```

Comment:
This is a very interesting example, and I'm glad you included it. There are two things I want to point out:

1. Unfortunately, some of the time this usage of shared locking followed by update is going to run into [lock promotion](https://github.com/cockroachdb/cockroach/issues/110435), which we don't support yet... makes it a little risky as an example.
2. Usually the way to avoid deadlocks is to always force locking to occur in a specific order.

To address both of these points, it might be worth adding a _third_ section after "Reserve row values using shared locks" (or as a subsection of "Reserve row values using shared locks") which shows how to properly avoid _both_ lock promotion and deadlock, by changing this statement from both transactions:

```sql
SELECT * FROM schedules
  WHERE day = '2023-12-05'
  FOR SHARE;
```

into this set of three statements:

```sql
-- shared locking of all doctors < this one
SELECT * FROM schedules
  WHERE day = '2023-12-05'
  AND doctor_id < foo
  ORDER BY doctor_id
  FOR SHARE;

-- exclusive locking of this doctor
SELECT * FROM schedules
  WHERE day = '2023-12-05'
  AND doctor_id = foo
  FOR UPDATE;

-- shared locking of all doctors > this one
SELECT * FROM schedules
  WHERE day = '2023-12-05'
  AND doctor_id > foo
  ORDER BY doctor_id
  FOR SHARE;
```

(where `foo` is the current doctor's ID).

But maybe this is too complicated for docs? Not sure. (You can see that simply using `ORDER BY doctor_id FOR UPDATE` for all rows is a simpler way to avoid deadlocks than this mixture of `FOR SHARE` and `FOR UPDATE`, so that's probably what I would recommend for a real customer.)

Reply:
I really appreciate this context! I do think it's a bit complicated for the docs, however. 

Since the example isn't showing the intended usage of FOR SHARE anyway (read rows without subsequently updating within the transaction), I tried simplifying the example. Do you think it works?

Reply:
I think it works. Very nice!